File: cpp/logging/src/Logging.cpp
/*
 * (c) 2017 Ionic Security Inc.
 * By using this code, I agree to the Terms & Conditions (https://www.ionic.com/terms-of-use/)
 * and the Privacy Policy (https://www.ionic.com/privacy-notice/).
 */

#include "ISAgent.h"
#include "ISChunkCrypto.h"
#include "ISAgentSDKError.h"
#include "ISLog.h"

#define MY_APPLICATION_CHANNEL_NAME "SampleLoggingApplication"

bool makeLinuxHomeDirPath(std::string pathFromHomeDir, std::string & fullPath);

void setupLogging() {
    //NOTE: Depending on the severity and channel name, this can cause many of the Ionic internal SDK actions to be logged.
    //	This is made available as it is useful for debugging, but you may want to filter on your channel instead
    //	of ISAGENT_LOG_CHANNEL to see content only from your application.
    ISLogFilterSeverity * pConsoleFilter = new ISLogFilterSeverity(SEV_INFO); // INFO and above goes to console
    ISLogWriterConsole * pConsoleWriter = new ISLogWriterConsole();
    pConsoleWriter->setFilter(pConsoleFilter);

    ISLogFilterSeverity * pFileFilter = new ISLogFilterSeverity(SEV_DEBUG); // DEBUG and above goes to file
    ISLogWriterFile * pFileWriter = new ISLogWriterFile("sample.log");
    pFileWriter->setFilter(pFileFilter);

    // Initialize log sink(s)
    ISLogSink * pSink = new ISLogSink();
    pSink->registerChannelName(ISAGENT_LOG_CHANNEL);
    pSink->registerChannelName(MY_APPLICATION_CHANNEL_NAME);
    pSink->registerWriter(pConsoleWriter);
    pSink->registerWriter(pFileWriter);

    // Initialize logger. We pass true to the constructor so that it owns all
    // the registered objects (the objects above that we created on the heap).
    ISLogImpl * pLogger = new ISLogImpl(true);
    pLogger->registerSink(pSink);

    // Assign logger to static interface ISLog
    ISLog::setSingleton(pLogger);
}

int main()
{
    std::string plainText = "Ionic Hello World";

    setupLogging();

    // Setup an agent object to talk to Ionic
    ISAgent agent;
    int nErrorCode;
#if __linux__
    //NOTE: On Linux, you must add additional code here, see "Getting Started" for C++ on Linux.
    std::string plainSepPath;
    if (!makeLinuxHomeDirPath(".ionicsecurity/profiles.pt", plainSepPath)) { // Makes the absolute path for ~/.ionic/profiles.pt
        std::cerr << "Error getting home directory path." << std::endl;
        return -1;
    }
    ISAgentDeviceProfilePersistorPlaintext plainPersistor;
    plainPersistor.setFilePath(plainSepPath);
    std::cout << "Initializing agent..." << std::endl;
    nErrorCode = agent.initialize(plainPersistor);
    if (nErrorCode != ISAGENT_OK) {
        std::cerr << "Error initializing agent: " << ISAgentSDKError::getErrorCodeString(nErrorCode) << std::endl;
    } else {
        std::cout << "A plaintext SEP was loaded from " << plainSepPath << std::endl;
    }
#else
    nErrorCode = agent.initialize();
    if (nErrorCode != ISAGENT_OK) {
        std::cerr << "Error initializing agent: " << ISAgentSDKError::getErrorCodeString(nErrorCode) << std::endl;
    }
#endif

    // Check if there are profiles.
    if (!agent.hasAnyProfiles()) {
        std::cout << "There are no device profiles on this device." << std::endl;
        std::cout << "Register a device before continuing." << std::endl;
        return -1;
    }

    // Setup a Chunk Crypto object to handle Ionic encryption
    ISChunkCryptoCipherAuto chunkCrypto(agent);

    // Encrypt the string using an Ionic-managed Key
    std::string encryptedText;
    int nErrorCode = chunkCrypto.encrypt(plainText, encryptedText);

    // Validate the response
    if (nErrorCode != ISCRYPTO_OK)
    {
        //std::cerr << "Error encrypting: " << ISAgentSDKError::getErrorCodeString(nErrorCode) << std::endl;
        ISLOGF_ERROR(MY_APPLICATION_CHANNEL_NAME, "Error encrypting: %s\n", ISAgentSDKError::getErrorCodeString(nErrorCode).c_str());
    }
    else
    {
        std::cout << "Plain Text: " << plainText << std::endl;
        std::cout << "Chunk-Encrypted String: " << encryptedText << std::endl;
    }
    ISLOGF_INFO(MY_APPLICATION_CHANNEL_NAME, "Press return to exit.");
    std::getchar();
    return 0;
}

bool makeLinuxHomeDirPath(std::string pathFromHomeDir, std::string & fullPath) {
    char const* tmp = getenv("HOME");
    if (tmp == NULL) {
        return false;
    } else {
        fullPath = std::string(tmp) + "/" + pathFromHomeDir;
        return true;
    }
}

File: java-jni/logging/src/main/java/com/ionic/samples/Logging.java
/*
 * (c) 2017 Ionic Security Inc.
 * By using this code, I agree to the Terms & Conditions (https://dev.ionic.com/use.html)
 * and the Privacy Policy (https://www.ionic.com/privacy-notice/).
 */

import com.ionicsecurity.sdk.Agent;
import com.ionicsecurity.sdk.AgentSdk;
import com.ionicsecurity.sdk.DeviceProfilePersistorPlainText;
import com.ionicsecurity.sdk.ChunkCipherAuto;
import com.ionicsecurity.sdk.KeyAttributesMap;
import com.ionicsecurity.sdk.Log;
import com.ionicsecurity.sdk.LogBase;
import com.ionicsecurity.sdk.LogFactory;
import java.util.List;
import java.util.ArrayList;

public class Logging
{
  public static void main(String[] args)
  {

    String plainText = "Hello, Ionic!";

    // Setup an agent object to talk to Ionic
    AgentSdk.initialize(null);  
    Agent agent = new Agent();

    if (System.getProperty("os.name").startsWith("Linux")) {
      DeviceProfilePersistorPlainText ptPersistor = new DeviceProfilePersistorPlainText();
      String sProfilePath = System.getProperty("user.home") + "/.ionicsecurity/profiles.pt";
      ptPersistor.setFilePath(sProfilePath);
      agent.initialize(ptPersistor);
    } else {
      agent.initialize();
    }

    // The configuration file, instead of programmatic commands, can be used to setup the logging.
    LogBase logger = LogFactory.getInstance().createFromConfigFile("logConfig.json");
    System.out.println(logger);
    Log.setSingleton(logger);
    Log.tc(Logging.class, "main", "LogDemoChannel", "This is a log of severity TRACE");
    Log.dc(Logging.class, "main", "LogDemoChannel", "This is a log of severity DEBUG");
    Log.ic(Logging.class, "main", "LogDemoChannel", "This is a log of severity INFO");
    Log.wc(Logging.class, "main", "LogDemoChannel", "This is a log of severity WARN");
    Log.ec(Logging.class, "main", "LogDemoChannel", "This is a log of severity ERROR");
    Log.fc(Logging.class, "main", "LogDemoChannel", "This is a log of severity FATAL");
    
    // Encrypt a string using the chunk data format.  
    KeyAttributesMap attrMap = new KeyAttributesMap();
    
    List<String> values = new ArrayList<String>();
    values.add("restricted"); //make sure you have access to keys marked as restricted before setting this key attribute
    attrMap.set("classification", values);
    ChunkCipherAuto chunkCrypto = new ChunkCipherAuto(agent);
    String encryptedText = chunkCrypto.encrypt(plainText, null);

    System.out.println("Plain Text: " + plainText);
    System.out.println("Chunk-Encrypted String: " + encryptedText);
  }
}
